sdk/src/main/scala/coop/rchain/sdk/dag/merging/DagMergingLogic.scala

```diff
@@ -0,0 +1,378 @@
+package coop.rchain.sdk.dag.merging
+
+import cats.Order
+import cats.syntax.all._
+import coop.rchain.sdk.dag.merging.data.Merge
+
+import scala.collection.compat.immutable.LazyList
+import scala.collection.immutable.Map
+import scala.collection.mutable
+import scala.math.Numeric.LongIsIntegral
+
+object DagMergingLogic {
+
+  /** All items in dependency chains. */
+  def withDependencies[D](of: Set[D], dependencyMap: Map[D, Set[D]]): Set[D] = {
+    def next(curOpt: Option[Set[D]]): Option[Set[D]] = curOpt.flatMap { c =>
+      val n = c.flatMap(dependencyMap.getOrElse(_, Set()))
+      n.nonEmpty.guard[Option].as(n)
+    }
+    LazyList.iterate(of.some)(next).takeWhile(_.nonEmpty).flatten.flatten.toSet
+  }
+
+  /** Deploys incompatible with finalized body. */
+  def incompatibleWithFinal[D](
+      acceptedFinally: Set[D],
+      rejectedFinally: Set[D],
+      conflictsMap: Map[D, Set[D]],
+      dependencyMap: Map[D, Set[D]]
+  ): Set[D] =
+    acceptedFinally.flatMap(conflictsMap) ++ rejectedFinally.flatMap(dependencyMap)
+
+  /** Split the scope into non overlapping partitions, greedily allocating intersecting chunks to bigger view. */
+  def partitionScope[D: Ordering](views: Seq[Set[D]]): Seq[Set[D]] = {
+    val r = LazyList.unfold(views) {
+      case (head @ seen) +: tail => (head, tail.map(_ -- seen)).some
+      case Seq()                 => none[(Set[D], List[Set[D]])]
+    }
+    r.toList
+  }
+
+  def computeRelationMap[D](
+      targetSet: Set[D],
+      sourceSet: Set[D],
+      relation: (D, D) => Boolean,
+      directed: Boolean
+  ): Map[D, Set[D]] =
+    targetSet.iterator
+      .flatMap(t => sourceSet.map((t, _)))
+      .foldLeft(mutable.Map.empty[D, Set[D]]) {
+        case (acc, (target, source)) =>
+          if (relation(target, source) && target != source) {
+            acc.update(source, acc.get(source).map(_ + target).getOrElse(Set(target)))
+            if (!directed)
+              acc.update(target, acc.get(target).map(_ + source).getOrElse(Set(source)))
+          }
+          acc
+      }
+      .toMap
+
+  /** Build conflicts map. */
+  def computeConflictsMap[D](
+      setA: Set[D],
+      setB: Set[D],
+      conflicts: (D, D) => Boolean
+  ): Map[D, Set[D]] =
+    computeRelationMap(setA, setB, conflicts, false)
+
+  /**
+    * Build dependency map.
+    * @param targetSet items that might depend on source set
+    * @param sourceSet items that target set might depend on
+    * @param depends dependency predicate
+    */
+  def computeDependencyMap[D](
+      targetSet: Set[D],
+      sourceSet: Set[D],
+      depends: (D, D) => Boolean
+  ): Map[D, Set[D]] =
+    computeRelationMap(targetSet, sourceSet, depends, true)
```

Comment:
This is typical partial application of a function if Boolean parameter is the first one. In that case these two functions can be defined as partially applying this first parameter.
```scala
def computeConflictsMap[D] =
  computeRelationMap[D](directed = false) _

def computeDependencyMap[D] =
  computeRelationMap[D](directed = true) _
```